### PR TITLE
Improve HCAD cold start

### PR DESCRIPTION
**Pull request by kaituo**
### Description

Kudos to Sudipto who designed the improvement.

ΔT is the length of the interval in minutes. A few tenets:

*If ΔT is small then we can wait for real data to fill up to 32 + shingleSize
*if ΔT is large then either user has a lot of data (and we can probe it slowly, or have large clusters) or the only option is to wait ...
*if ΔT divides 1440 then there is a meaning to a “day” — this is a likely common case.
*Current solution in prod uses linear interpolation — it will take significant work in changing that. Given that, it makes sense to embrace linearity all the way for missing data as well.
*If we have 32 + shingleSize (hopefully recent) values, RCF can get up and running. It will be noisy — there is a reason that default size is 256 (+ shingle size), but it may be more useful for people to start seeing some results.

We have two parameters : numberOfSamples and strideLength —

*We probe numberOfSamples + 1 values at strideLength * ΔT gap
*For each consecutive value present we interpolate into strideLength number of pieces. 0, as returned by the engine should constitute a valid answer, “null” is a missing answer — it may be that 0 is meaningless in some cases, but 0 is also meaningful in some cases. It may be that the query defining the metric is ill-formed, but that cannot be solved by the cold-start strategy of the AD plugin — if we attempt to do that, we will have issues with legitimate interpretations of 0.
*For the missing entries we use linear interpolation as well. Denote the Samples S0, S1, ... as samples in reverse order of time. Each [Si,Si−1]corresponds to strideLength * ΔTgap. If we get samples for S0, S1, S4 (both S2 and S3 are missing) then we interpolate the [S4,S1] into 3*strideLength pieces.
*If the above provides (32+shingleSize) points (note that if S0 is missing or all Sif or some i > N is missing then we would miss a lot of points — but the points we will get are contiguous based on the suggestion) then we have a model. Otherwise we issue another round of query — if there is any sample in the second round then we would have 32 + shingleSize points. If there is no sample in the second round then we should wait for real data.
*If there is no data — there is ultimately nothing that can be done.

How to set numberOfSamples and strideLength?
*Suppose ΔT≤30 and divides 60. Then set numberOfSamples = ceil ( (shingleSize + 32)/ 24 )*24 and strideLength = 60/ΔT. Note that if there is enough data — we may have a lot more than shingleSize+32 points — which is only good.
*Set numberOfSamples = (shingleSize + 32) and strideLength = 1. This should be an uncommon case, but if someone wants a 23 minutes interval — and the system permits -- let's give it to them. Note the smallest ΔT that does not divide 60 is 7 which is quite large to wait for one data point.

Testing done:
1. added precision tests and various unit tests to cover changes.
2. Manually verified HCAD cold start does not break.
 
### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/ad/NodeState.java [hunk 1]
```diff
@@ -166,6 +170,24 @@ public void setColdStartRunning(boolean coldStartRunning) {
         refreshLastUpdateTime();
     }
 
+    /**
+    *
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/NodeState.java:174]: fixed

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java [hunk 2]
```diff
@@ -355,6 +377,8 @@ private void trainModelFromDataSegments(
 
         // save to checkpoint
         checkpointWriteQueue.write(entityState, true, RequestPriority.MEDIUM);
+        // clear samples after using
+        dataPoints.clear();
```
[the remark this answers is not in the document]
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:381]: good catch.  Removed.

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java [hunk 3]
```diff
@@ -263,8 +263,6 @@ private void coldStart(
                             checkpointWriteQueue.write(modelState, true, RequestPriority.MEDIUM);
                             logger.info("Not enough data to train entity: {}, currently we have {}", modelId, samples.size());
                         }
-                        // save to checkpoint
-                        checkpointWriteQueue.write(modelState, true, RequestPriority.MEDIUM);
```
[the remark this answers is not in the document]
Comment (wnbts) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java]: ok. that makes sense.

**Review by wnbts — APPROVED** [review]

File: src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java [hunk 4]
```diff
@@ -1000,9 +991,9 @@ public void getColdStartSamplesForPeriods(
                         .stream()
                         .filter(InternalDateRange.class::isInstance)
                         .flatMap(agg -> ((InternalDateRange) agg).getBuckets().stream())
-                        .filter(bucket -> bucket.getFrom() != null)
+                        .filter(bucket -> bucket.getFrom() != null && bucket.getFrom() instanceof ZonedDateTime)
                         .filter(bucket -> bucket.getDocCount() > docCountThreshold)
-                        .sorted(Comparator.comparing((Bucket bucket) -> Long.valueOf(bucket.getFromAsString())))
+                        .sorted(Comparator.comparing((Bucket bucket) -> (ZonedDateTime) bucket.getFrom()))
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/feature/SearchFeatureDao.java:996]: Got it.

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java [hunk 5]
```diff
@@ -476,27 +456,176 @@ private void getEntityColdStartData(String detectorId, Entity entity, ActionList
...
+     * but if someone wants 23 minutes interval — and the system permits lets give it to them. Note the
+     * smallest delta that does not divide 60 is 7 which is quite large to wait for one data point.
+     * @return the chosen strideLength and numberOfSamples
+     */
+    private Pair<Integer, Integer> selectRangeParam(AnomalyDetector detector) {
+        long delta = detector.getDetectorIntervalInMinutes();
+        int shingleSize = detector.getShingleSize();
+        int strideLength = defaulStrideLength;
+        int numberOfSamples = defaultNumberOfSamples;
+        if (delta <= 30 && 60 % delta == 0) {
```
Comment (sudiptoguha) [inline src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:600]: if  (delta == 1) then the % operator fails I think (even though 1 divides 60) ...

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java [hunk 6]
```diff
@@ -243,13 +252,17 @@ private void coldStart(
...
-                        if (hasEnoughSample(dataPoints, modelState) == false) {
-                            combineTrainSamples(dataPoints, modelId, modelState);
+                        if (samples.size() >= numMinSamples) {
+                            trainModelFromDataSegments(samples, entity, modelState, detector.getShingleSize());
+                            logger.info("Succeeded in training entity: {}", modelId);
                         } else {
-                            trainModelFromDataSegments(dataPoints, entity, modelState, detector.getShingleSize());
+                            // save to checkpoint
+                            checkpointWriteQueue.write(modelState, true, RequestPriority.MEDIUM);
+                            logger.info("Not enough data to train entity: {}, currently we have {}", modelId, samples.size());
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:266]: ok

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java (hunk 5 above)
Comment (kaituo) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:600]: jshell> 60 % 2
$1 ==> 0
jshell> 60 % 1
$2 ==> 0
jshell> 60 % 7
$3 ==> 4

It seems 1 works.

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java [hunk 7]
```diff
@@ -476,27 +456,176 @@ private void getEntityColdStartData(String detectorId, Entity entity, ActionList
...
+                if (featuresOptional.isPresent()) {
+                    // we only need the most recent two samples
+                    // For the missing samples we use linear interpolation as well.
+                    // Denote the Samples S0, S1, ... as samples in reverse order of time.
+                    // Each [Si​,Si−1​]corresponds to strideLength * detector interval.
+                    // If we got samples for S0, S1, S4 (both S2 and S3 are missing), then
+                    // we interpolate the [S4,S1] into 3*strideLength pieces.
+                    if (lastSample != null) {
+                        // right sample has index i and feature featuresOptional.get()
+                        int numInterpolants = (i - lastSample.getLeft()) * stride + 1;
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:503]: ok

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java [hunk 8]
```diff
@@ -575,49 +692,41 @@ public void trainModelFromExistingSamples(ModelState<EntityModel> modelState, in
...
+     * combining them without reordering based on time stamps is not ok. We might introduce
+     * anomalies in the process.
+     * An ideal solution would be to record time stamps of data points and combine existing
+     * samples and cold start samples and do interpolation afterwards. Recording time stamps
+     * requires changes across the board like bwc in checkpoints. A pragmatic solution is to use
+     * job enabled time as the end time of cold start period as it is easier to combine
+     * existing samples with cold start data. We just need to appends existing samples after
+     * cold start data as existing samples all happen after job enabled time. There might
+     * be some gaps in between the last cold start sample and the first accumulated sample.
+     * We will need to accept that precision loss in current solution.
```
[the remark this answers is not in the document]
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:709]: Ok - we should get back to this at some point.

File: src/main/java/org/opensearch/ad/ml/EntityColdStarter.java (hunk 5 above)
Comment (sudiptoguha) [reply src/main/java/org/opensearch/ad/ml/EntityColdStarter.java:600]: ok

**Review by sudiptoguha — APPROVED** [review]

**Pull request merged**